perf/perf.go

```diff
@@ -0,0 +1,520 @@
+	case unix.EPERM:
+		return -1, errors.WithMessage(unix.EPERM, "insufficient capability to request exclusive access")
+	case unix.ESRCH:
+		return -1, errors.WithMessage(unix.ESRCH, "pid does not exist")
+	default:
+		return -1, err
+	}
+}
+
+func addToEpoll(epollfd, fd int, data, pad int32) error {
+	// The representation of EpollEvent isn't entirely accurate.
+	// According to man 2 epoll_wait, Fd and Pad are really a union,
+	// and can be used to store arbitrary values, not just fds.
+	event := unix.EpollEvent{
+		Events: unix.EPOLLIN,
+		Fd:     data,
+		Pad:    pad,
+	}
+
+	err := unix.EpollCtl(epollfd, unix.EPOLL_CTL_ADD, fd, &event)
+	return errors.Wrap(err, "can't add fd to epoll")
+}
+
+func (ring *perfEventRing) Close() {
+	runtime.SetFinalizer(ring, nil)
+	unix.Close(ring.fd)
+	unix.Munmap(ring.mmap)
+
+	ring.fd = -1
+	ring.mmap = nil
+}
+
+func (ring *perfEventRing) LoadHead() {
+	ring.reader.loadHead()
+}
+
+// NB: Has to be preceded by a call to LoadHead.
+func (ring *perfEventRing) ReadRecord() (*Record, error) {
+	if ring.fd == -1 {
+		return nil, errors.New("closed")
+	}
+
+	defer ring.reader.writeTail()
+	return readRecord(ring.reader)
+}
+
+// Record contains either a sample or a counter of the
+// number of lost samples.
+type Record struct {
+	// The data submitted via bpf_perf_event_output.
+	// They are padded with 0 to have a 64-bit alignment.
+	// If you are using variable length samples you need to take
+	// this into account.
+	RawSample []byte
+
+	// The number of samples which could not be output, since
+	// the ring buffer was full.
+	LostSamples uint64
+}
+
+func readRecord(rd io.Reader) (*Record, error) {
+	const (
+		perfRecordLost   = 2
+		perfRecordSample = 9
+	)
+
+	var header perfEventHeader
+	err := binary.Read(rd, internal.NativeEndian, &header)
+	if err == io.EOF {
+		return nil, nil
+	}
+
+	if err != nil {
+		return nil, errors.Wrap(err, "can't read event header")
+	}
+
+	switch header.Type {
+	case perfRecordLost:
+		lost, err := readLostRecords(rd)
+		if err != nil {
+			return nil, err
+		}
+
+		return &Record{LostSamples: lost}, nil
+
+	case perfRecordSample:
+		raw, err := readRawSample(rd)
+		if err != nil {
+			return nil, err
+		}
+
+		return &Record{RawSample: raw}, nil
+
+	default:
+		return nil, errors.Errorf("unknown event type %d", header.Type)
+	}
+}
+
+func readLostRecords(rd io.Reader) (uint64, error) {
+	var lostHeader struct {
+		ID   uint64
+		Lost uint64
+	}
+
+	err := binary.Read(rd, internal.NativeEndian, &lostHeader)
+	if err != nil {
+		return 0, errors.Wrap(err, "can't read lost records header")
+	}
+
+	return lostHeader.Lost, nil
+}
+
+func readRawSample(rd io.Reader) ([]byte, error) {
+	var size uint32
+	if err := binary.Read(rd, internal.NativeEndian, &size); err != nil {
+		return nil, errors.Wrap(err, "can't read sample size")
+	}
+
+	data := make([]byte, int(size))
+	_, err := io.ReadFull(rd, data)
```

Comment:
I realise there may actually be some ugly CGo hacks behind Cilium's ability to retain / pass pointers into the ringbuffer here, not sure exactly what it looks like for a variable-length golang reference to be passed back out to the caller, in a safe manner, without allocations, when the sample structure isn't actually known by the library...

Reply:
> These above two lines I'm wondering about from a high level, it's an enforced memory allocation/copy of the data from the ringbuffer, for every single sample.

I kind of figured you would raise this :) It was the reason I re-wrote the Reader to not use a buffered channel. How about something like this:

`ReadRecord(*Record) error`

You can then pre-allocate Record.RawSample to your liking, and afterwards unmarshal into your structure.

I didn't add it to this PR since it begs the question what happens when RawSample is to small. Return an error? Discard and re-allocate?

> Is it worth reconsidering the Read() API to use a model that allows arbitrary user code to interact with the memory-mapped buffer directly?

See my preferred solution above.

Accessing the ring buffer as []byte doesn't work because of records wrapping around the buffer. You need an `io.Reader` to abstract that away. Once you have a reader you kind of need to do a copy to get a linear buffer anyway, which I suspect is why your current code does a copy as well. Once you have the linear buffer you can unmarshal from it.

Reply:
I've added a `ReadRecord` to the PR, please take a look. Usage would work like this:

```go
var rec Record
for {
  reader.ReadRecord(&rec)
  // ...
}
```

This will successively grow the buffer to the maximum sample you receive.

Reply:
When I think a bit more, I don't know if it's worth adding such an API at this point. If you're happy with this new version you came up with, and the testing / corner cases are all handled, then we can move forward with it, but otherwise maybe we can defer it until we get a strong argument for why we need to do it. It's certainly a bit more ugly from a go abstraction point of view.

(LMK, I can take a closer look, didn't get a chance today)

Reply:
OK, removed ReadRecord again. Let's tackle reducing allocations in a follow up.